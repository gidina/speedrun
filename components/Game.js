import React from "react";
import PropTypes from 'prop-types';
import Link from 'next/link';

import "./index.css";

const Game = props => {
    const { game } = props;
    return (
        <li className="game">
            <img alt="" className="game-img" src={game.logoUrl} />
            <Link as={`/game/${game.id}`} href={`/game?id=${game.id}`}>
                <a className="game-title">{game.name}</a>
            </Link>
        </li>
    );
};

export default Game;

Game.propTypes = {
    game: PropTypes.shape({
        id: PropTypes.integer,
        logoUrl: PropTypes.string,
        name: PropTypes.string
    }).isRequired
};