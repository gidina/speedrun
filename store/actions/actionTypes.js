export const FETCH_GAMES_SUCCESS = 'FETCH_GAMES_SUCCESS';
export const FETCH_GAMES_ERROR = 'FETCH_GAMES_ERROR';

export const FETCH_RUNS_SUCCESS = 'FETCH_RUNS_SUCCESS';
export const FETCH_RUNS_ERROR = 'FETCH_RUNS_ERROR';