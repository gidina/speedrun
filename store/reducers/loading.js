import { 
    FETCH_GAMES_REQUEST, 
    FETCH_GAMES_SUCCESS,
    FETCH_GAMES_ERROR,
    FETCH_RUNS_REQUEST,
    FETCH_RUNS_SUCCESS,
    FETCH_RUNS_ERROR
} from "../actions/actionTypes";

const initialState = {
    loadingGames: false,
    loadingRuns: false
};

const loading = (state = initialState, action) => {
    switch (action.type) {
        case FETCH_GAMES_REQUEST:
            return {
                ...state,
                loadingGames: true
            };
        case FETCH_GAMES_SUCCESS:
        case FETCH_GAMES_ERROR:
            return {
                ...state,
                loadingGames: false
            };
        case FETCH_RUNS_SUCCESS:
        case FETCH_RUNS_ERROR:
            return {
                ...state,
                loadingRuns: false
            };
        case FETCH_RUNS_REQUEST:
            return {
                ...state,
                loadingRuns: true
            };
        default:
            return state;
    }
};
  
export default loading;

export const isFetchingGames = state => state.loadingGames;
export const isFetchingRuns = state => state.loadingRuns;