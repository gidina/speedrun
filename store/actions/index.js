import fetch from 'isomorphic-unfetch';
import { FETCH_GAMES_SUCCESS, FETCH_GAMES_ERROR } from "./actionTypes";

const fetchGamesSuccess = games => ({ type: FETCH_GAMES_SUCCESS, games });
const fetchGamesError = error => ({ type: FETCH_GAMES_ERROR, error });

export const fetchGames = () => {
    return async (dispatch, getState) => {
        
        const games = getState().games;
        if (games.length !== 0) return;
        
        try {
            const res = await fetch('https://www.speedrun.com/api/v1/games');
            const result = await res.json();
            const games = result.data;
            dispatch(fetchGamesSuccess(games));
        } catch(err) {
            dispatch(fetchGamesError(err));
        }
    };
};