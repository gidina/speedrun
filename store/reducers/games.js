import { FETCH_GAMES_SUCCESS, FETCH_GAMES_ERROR } from "../actions/actionTypes";

const initialState = [];
  
const games = (state = initialState, action) => {
    switch (action.type) {
        case FETCH_GAMES_SUCCESS:
            return action.games;
        case FETCH_GAMES_ERROR:
            return action.error;
        default:
            return state;
    }
};
  
export default games;

const findByID = (obj, id) => obj.id === id;
const findByRel = (obj, rel) => obj.rel === rel;
export const getGames = state => state;
export const getGameById = (state, id) => {
    return state.find((game) => findByID(game, id))
};
export const areGamesLoaded = state => state.length !== 0;
export const getRunsUrl = (state, id) => {
    const game = getGameById(state, id);

    if (!game) return;

    const link = game.links.find(link => findByRel(link, 'runs'));
    
    if (!link) return;
    
    return link.uri;
};