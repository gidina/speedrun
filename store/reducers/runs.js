import { FETCH_RUNS_SUCCESS, FETCH_RUNS_ERROR } from "../actions/actionTypes";

const initialState = {};
  
const runs = (state = initialState, action) => {
    switch (action.type) {
        case FETCH_RUNS_SUCCESS: {
            const { id, runs } = action;
            return {
                ...state,
                [id]: runs
            }
        } 
        case FETCH_RUNS_ERROR:
            return action.error;
        default:
            return state;
    }
};
  
export default runs;

export const getFirstRunByGameId = (state, id) => {
    const gameRuns = state[id];
    if (!gameRuns) return;

    return gameRuns[0];
};
export const areRunsLoadedByGameId = (state, id) => !!state[id];