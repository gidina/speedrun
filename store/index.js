import { createStore , applyMiddleware} from 'redux';
import reducers from "./reducers";
import thunkMiddleware from 'redux-thunk';

export const initializeStore = (initialState) => {
    return createStore(reducers, initialState, applyMiddleware(thunkMiddleware));
};