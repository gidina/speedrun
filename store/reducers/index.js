import { combineReducers } from 'redux';
import * as gamesReducer from "./games";
import * as runsReducer from "./runs";
import * as loadingReducer from "./loading";

const rootReducer = combineReducers({
    games: gamesReducer.default,
    runs: runsReducer.default,
    loading: loadingReducer.default
});

export default rootReducer;

export const getGames = state => gamesReducer.getGames(state.games);
export const areGamesLoaded = state => gamesReducer.areGamesLoaded(state.games);
export const getRunsUrl = (state, id) => gamesReducer.getRunsUrl(state.games, id);
export const getGameById = (state, id) => gamesReducer.getGameById(state.games, id);
export const getSelectedGame = (state, id) => {
    const firstRun = runsReducer.getFirstRun(state.runs);
    const game = gamesReducer.getGameById(state.games, id);

    if (!firstRun || !game) {
        return;
    }

    return {
        id,
        name: game.names.international,
        logoUrl: game.assets['cover-small'].uri,
        firstRun
    }
};
export const isFetchingGames = state => loadingReducer.isFetchingGames(state.loading);
export const isFetchingRuns = state => loadingReducer.isFetchingRuns(state.loading);