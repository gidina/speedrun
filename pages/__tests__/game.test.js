import React from "react";
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import Game from "../game";
import { getSelectedGame } from "../../store/reducers";
import { onGameDetailPageEnter } from "../../store/actions";

jest.mock("../../components/GameDetail", () => "GameDetail-mock");
jest.mock("../../store/reducers/index");
jest.mock("../../store/actions");

const selectedGameMock = {
    id: "game-1", 
    name: "international-name",
    logoUrl: "cover-small-uri", 
    firstRun: {
        firstPlayerName: "player-name",
        timePlayed: 23,
        videoUrl: "link-uri"
    }
};

getSelectedGame.mockReturnValue(selectedGameMock);
onGameDetailPageEnter.mockReturnValue({ type: "DUMMY_FETCH_GAMES" });

const initialState = {};

const mockStore = configureStore();
const store = mockStore(initialState);

beforeEach(() => {
    jest.clearAllMocks();
    store.clearActions();
});

it('Renders correctly', () => {
    const wrapper = <Provider store={store}><Game /></Provider>;
    const tree = renderer
        .create(wrapper)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('#getInitialProps', async () => {
    const gameId = "game-1";
    const context = {
        reduxStore: store,
        query: {
            id: gameId
        } 
    };
    await Game.getInitialProps(context);
    expect(onGameDetailPageEnter).toHaveBeenCalledTimes(1);
    expect(onGameDetailPageEnter).toHaveBeenCalledWith(gameId); 
    expect(store.getActions()).toContainEqual({ type: "DUMMY_FETCH_GAMES" });
});
