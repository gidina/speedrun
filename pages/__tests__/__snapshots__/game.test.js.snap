// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<GameDetail-mock
  dispatch={[Function]}
  selectedGame={
    Object {
      "firstRun": Object {
        "firstPlayerName": "player-name",
        "timePlayed": 23,
        "videoUrl": "link-uri",
      },
      "id": "game-1",
      "logoUrl": "cover-small-uri",
      "name": "international-name",
    }
  }
/>
`;
